File: workspace/src/my_py_pkg/my_py_pkg/publisher_with_params.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from example_interfaces.msg import String

class PublisherWithParams(Node):
    """ Publisher example that periodically sends out a string """

    def __init__(self):
        super().__init__("publisher_with_params")

        self.declare_parameter("message", "Hello")
        self.declare_parameter("timer_period", 1.0)

        self._message = self.get_parameter("message").value
        self._timer_period = self.get_parameter("timer_period").value

        self._publisher = self.create_publisher(String, "my_topic", 10)
        self._timer = self.create_timer(self._timer_period, self._timer_callback)

        self.add_post_set_parameters_callback(self._post_parameters_callback)

    def _post_parameters_callback(self, params: list[rclpy.Parameter]):
        """ Set parameters after node started """

        for param in params:
            if param.name == "message":
                # Update `message` Parameter
                self._message = param.value
                self.get_logger().info(f"Set {param.name} to {param.value}")

            elif param.name == "timer_period":
                # Update `timer_period` Parameter
                self._timer_period = param.value
                self.get_logger().info(f"Set {param.name} to {param.value}")

                self._timer.cancel()
                self._timer = self.create_timer(self._timer_period, self._timer_callback)
            
            else:
                self.get_logger().warn(f"Unknown parameter: {param.name}")

    def _timer_callback(self):
        """ Publishes a simple message to topic """

        msg = String()
        msg.data = self._message

        self._publisher.publish(msg)
        self.get_logger().info(f"Publishing: {msg.data}")

def main(args = None):
    """ Main entrypoint """

    node = None
    try:
        rclpy.init()
        node = PublisherWithParams()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: workspace/src/my_cpp_pkg/src/publisher_with_params.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class PublisherWithParams : public rclcpp::Node
{

public:
  
  PublisherWithParams() : Node{"publisher_with_params"} // Call the parent constructor from the initilizer list 
  {
    this->declare_parameter("message", "Hello");
    this->declare_parameter("timer_period", 1.0);

    message_ = this->get_parameter("message").as_string();
    timer_period_ = this->get_parameter("timer_period").as_double();
    
    param_cb_handle_ = this->add_post_set_parameters_callback(std::bind(&PublisherWithParams::post_parameters_callback, this, std::placeholders::_1));

    publisher_ = this->create_publisher<example_interfaces::msg::String>("my_topic", 10);
    timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period_), std::bind(&PublisherWithParams::timer_callback, this));

  }

private:
  // Member varibles
  rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::node_interfaces::PostSetParametersCallbackHandle::SharedPtr param_cb_handle_;
  std::string message_;
  double timer_period_;
  
  // Private methods
  void post_parameters_callback(const std::vector<rclcpp::Parameter> &parameters) {
    
    for (const auto &param : parameters) {
        if (param.get_name() == "message") {
          message_ = param.as_string();
          RCLCPP_INFO(this->get_logger(), "Set %s to %s", param.get_name().c_str(), param.as_string().c_str());
        }
        else if (param.get_name() == "timer_period"){
          timer_period_ = param.as_double();
          RCLCPP_INFO(this->get_logger(), "Set %s to %.2lf", param.get_name().c_str(), param.as_double());

          timer_->cancel();
          timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period_), std::bind(&PublisherWithParams::timer_callback, this));
        }
        else {
          RCLCPP_WARN(this->get_logger(), "Unknown parameter: %s", param.get_name().c_str());
        }
    }

  }

  void timer_callback() {
    
    auto msg = example_interfaces::msg::String();
    msg.data = message_;

    publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Publishing: %s", msg.data.c_str());

  }

};

// Entry point
int main(int argc, char* argv[]) {
  
  rclcpp::init(argc, argv);

  auto node = std::make_shared<PublisherWithParams>();
  rclcpp::spin(node);

  rclcpp::shutdown();

  return 0;
}